lib/charms/operator_libs_linux/v0/juju_systemd_notices.py

```diff
@@ -0,0 +1,476 @@
+        # start back up if the underlying machine is rebooted.
+        self._systemd_notices.stop()
+
+    def _on_slurmd_started(self, _: ServiceStartedEvent) -> None:
+        self.unit.status = ActiveStatus()
+        time.sleep(60)
+
+        # This will trigger the juju-systemd-notices daemon to
+        # emit a `service-slurmd-stopped` event.
+        systemd.service_stop("slurmd")
+
+    def _on_slurmd_stopped(self, _: ServiceStoppedEvent) -> None:
+        self.unit.status = BlockedStatus("slurmd not running")
+```
+"""
+
+__all__ = ["ServiceStartedEvent", "ServiceStoppedEvent", "SystemdNotices"]
+
+import argparse
+import asyncio
+import functools
+import logging
+import re
+import signal
+import subprocess
+import sys
+import textwrap
+from pathlib import Path
+from typing import List
+
+from dbus_fast.aio import MessageBus
+from dbus_fast.constants import BusType, MessageType
+from dbus_fast.errors import DBusError
+from dbus_fast.message import Message
+from ops.charm import CharmBase
+from ops.framework import EventBase
+
+LIBID = "HPCTEAMUSEONLY"
+LIBAPI = 0
+LIBPATCH = 1
+PYDEPS = ["dbus-fast>=1.90.2"]
+
+_logger = logging.getLogger(__name__)
+_juju_unit = None
+_service_states = {}
+_service_hook_regex_filter = re.compile(r"service-(?P<service>[\w\\:-]*)-(?:started|stopped)")
+_DBUS_CHAR_MAPPINGS = {
+    "_5f": "_",  # _ must be first since char mappings contain _.
+    "_40": "@",
+    "_2e": ".",
+    "_2d": "-",
+    "_5c": "\\",
+}
+
+
+def _systemctl(*args) -> None:
+    """Control systemd by via executed `systemctl ...` commands.
+
+    Raises:
+        subprocess.CalledProcessError: Raised if systemctl command fails.
+    """
+    cmd = ["systemctl", *args]
+    _logger.debug(f"systemd: Executing command {cmd}")
+    try:
+        subprocess.check_output(cmd)
+    except subprocess.CalledProcessError as e:
+        _logger.error(f"systemctl command failed: {e}")
+        raise
+
+
+_daemon_reload = functools.partial(_systemctl, "daemon-reload")
+_start_service = functools.partial(_systemctl, "start")
+_stop_service = functools.partial(_systemctl, "stop")
+_enable_service = functools.partial(_systemctl, "enable")
+_disable_service = functools.partial(_systemctl, "disable")
+
+
+class ServiceStartedEvent(EventBase):
+    """Event emitted when service has started."""
+
+
+class ServiceStoppedEvent(EventBase):
+    """Event emitted when service has stopped."""
+
+
+class SystemdNotices:
+    """Observe systemd services on your machine base."""
+
+    def __init__(self, charm: CharmBase, services: List[str]) -> None:
+        """Instantiate systemd notices service."""
+        self._charm = charm
+        self._services = services
+        unit_name = self._charm.unit.name.replace("/", "-")
+        self._service_file = Path(f"/etc/systemd/system/juju-{unit_name}-systemd-notices.service")
+
+        _logger.debug(f"Attaching systemd notice events to charm {self._charm.__class__.__name__}")
```

Comment:
```suggestion
        _logger.debug("Attaching systemd notice events to charm %s", self._charm.__class__.__name__)
```

Reply:
Refactored in commit https://github.com/canonical/operator-libs-linux/pull/108/commits/3bd603a65b70ec45dfbc4ceff82114739179f9c6